roles/mining-proxy/src/lib/downstream_mining.rs

```diff
@@ -483,44 +481,56 @@ impl
     }
 }
 
-use network_helpers_sv2::plain_connection_tokio::PlainConnection;
-use std::net::SocketAddr;
-use tokio::net::TcpListener;
-
-pub async fn listen_for_downstream_mining(address: SocketAddr) {
-    info!("Listening for downstream mining connections on {}", address);
-    let listner = TcpListener::bind(address).await.unwrap();
+pub async fn listen_for_downstream_mining(
+    listener: TcpListener,
+    mut shutdown_rx: TokioReceiver<()>,
+) {
     let mut ids = roles_logic_sv2::utils::Id::new();
-
-    while let Ok((stream, _)) = listner.accept().await {
-        let (receiver, sender): (Receiver<EitherFrame>, Sender<EitherFrame>) =
-            PlainConnection::new(stream).await;
-        let node = DownstreamMiningNode::new(receiver, sender, ids.next());
-
-        task::spawn(async move {
-            let mut incoming: StdFrame = node.receiver.recv().await.unwrap().try_into().unwrap();
-            let message_type = incoming.get_header().unwrap().msg_type();
-            let payload = incoming.payload();
-            let routing_logic = super::get_common_routing_logic();
-            let node = Arc::new(Mutex::new(node));
-
-            // Call handle_setup_connection or fail
-            match DownstreamMiningNode::handle_message_common(
-                node.clone(),
-                message_type,
-                payload,
-                routing_logic,
-            ) {
-                Ok(SendToCommon::RelayNewMessageToRemote(_, message)) => {
-                    let message = match message {
-                        roles_logic_sv2::parsers::CommonMessages::SetupConnectionSuccess(m) => m,
+    loop {
+        tokio::select! {
+            accept_result = listener.accept() => {
+                match accept_result {
+                    Ok((stream, _)) => {
+                let (receiver, sender): (Receiver<EitherFrame>, Sender<EitherFrame>) =
+                    PlainConnection::new(stream).await;
+                let node = DownstreamMiningNode::new(receiver, sender, ids.next());
+
+                task::spawn(async move {
```

Comment:
I wonder if this `task::spawn` is needed here as we are already in a `select!`

Reply:
Nice catch! We can remove this `task::spawn` since the `tokio::select!` block already provides an asynchronous context for handling operations concurrently.

Reply:
I've removed `spawn` and simplified the code a bit